### PR TITLE
[sycl-post-link] Add SYCLStripDeadDebugInfo pass

### PR DESCRIPTION
**Pull request by mlychkov**
StripDeadDebugInfo pass from IPO is not stripping debug info thoroughly.
Split modules may still contain a lot of useless debug info after this
pass. If input module has a lot of debug info then device code after
split may significantly grow due to many duplications.
If making strip rules more strict then device code after split may be
reduced by 2-10 times, but there is no 100% garantie that these changes
doesn't corrupt debug info in any case. So, changing StripDeadDebugInfo
pass directly may add regressions to non sycl-post-link components that
use this pass.
Add SYCLStripDeadDebugInfo pass based on StripDeadDebugInfo pass with
more strict rules. Using these rules is enabled by default. An option
'-safe-strip-debug-info' may be used to roll back to the stripping algorithm
realized in StripDeadDebugInfo pass.

Signed-off-by: Mikhail Lychkov <mikhail.lychkov@intel.com>

### CODE REVIEW

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 1]
```diff
@@ -197,6 +198,12 @@ cl::opt<bool> DeviceGlobals{
     cl::desc("Lower and generate information about device global variables"),
     cl::cat(PostLinkCat)};
 
+cl::opt<bool> AggressiveStripDebugInfo{
+    "aggressive-strip-debug-info",
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:202]:
```suggestion
    "aggressively-strip-dead-debug-info",
```

File: llvm/tools/sycl-post-link/SYCLStripDeadDebugInfo.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+  std::set<DIGlobalVariableExpression *> LiveGVs;
+  for (GlobalVariable &GV : M.globals()) {
+    SmallVector<DIGlobalVariableExpression *, 1> GVEs;
+    GV.getDebugInfo(GVEs);
+    for (auto *GVE : GVEs)
+      LiveGVs.insert(GVE);
+  }
+
+  std::unordered_set<const DISubprogram *> PresentFunctionsSPs;
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SYCLStripDeadDebugInfo.cpp:56]: It took me some time to figure out the exact change compared to the original version of the algorithm, could you please add an explicit comment highlighting this change?

My understanding is that the change is to expect some `DISubprogram` not being associated with any functions (due to functions being removed by device code split earlier) and therefore, not to mark corresponding compilation units as live, right?

File: llvm/tools/sycl-post-link/sycl-post-link.cpp (hunk 1 above)
Comment (mlychkov) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: The option name was changed.

File: llvm/tools/sycl-post-link/SYCLStripDeadDebugInfo.cpp (hunk 2 above)
Comment (mlychkov) [reply llvm/tools/sycl-post-link/SYCLStripDeadDebugInfo.cpp:56]: Comments were added.
You are right. First optimization is to skip compile units if they have no `DISubprogram`s associated with any function in current submodule.
Second proposed optimization is to skip compile units with constant global expressions that also have no `DISubprogram`s associated with any function in current submodule.

**Pull request closed without merging**